src/crimson/osd/backfill_state.cc

```diff
@@ -409,7 +409,35 @@ BackfillState::PrimaryScanning::react(PrimaryScanned evt)
   LOG_PREFIX(BackfillState::PrimaryScanning::react::PrimaryScanned);
   DEBUGDPP("", pg());
   backfill_state().backfill_info = std::move(evt.result);
-  return transit<Enqueuing>();
+  if (!backfill_state().next) {
+    return transit<Enqueuing>();
+  } else {
+    DEBUGDPP("backfill cancelled, not going Enqueuing", pg());
+    backfill_state().go_enqueuing_on_resume();
+  }
+  return discard_event();
```

Comment:
General question, when will we eventually clean a cancelled backfill operation? Do we have a limit to the time its being kept alive so it can be resumed?

Reply:
I think a cancelled backfill will eventually either resume or be dropped on pg interval changes. @athanatos Am I right?

Reply:
The event in PeeringState that eventually leads to this is actually called DeferBackfill, it gets called if the recovery reservation is revoked before it completes.  As part of that pathway, PeeringState (see Backfilling::react(const DeferBackfill&) immediately restarts the backfill reservation process.  I'd really like to avoid the word "canceled" anywhere here -- it really just means that we temporarily don't have a recovery IO reservation.  The on_backfill_cancelled() PeeringState::PeeringListener method should probably be renamed as well (on_backfill_suspended?), but that's out of scope for this PR.

As long as the up set for the interval includes osds that need to be backfilled, the primary will continue trying to do so.  Backfill will either complete or will be made irrelevant as part of an interval change -- being deferred only delays it.